packages/nx/src/command-line/import/import.ts

```diff
@@ -0,0 +1,270 @@
+import { readNxJson } from '../../config/nx-json';
+import { workspaceRoot } from '../../utils/workspace-root';
+import {
+  detectPackageManager,
+  getPackageManagerCommand,
+} from '../../utils/package-manager';
+import { resetWorkspaceContext } from '../../utils/workspace-context';
+import { runInstall } from '../init/implementation/utils';
+import { getBaseRef } from '../../utils/command-line-utils';
+import { prepareSourceRepo } from './utils/prepare-source-repo';
+import { mergeRemoteSource } from './utils/merge-remote-source';
+import {
+  getPackagesInPackageManagerWorkspace,
+  needsInstall,
+} from './utils/needs-install';
+
+const importRemoteName = '__tmp_nx_import__';
+
+export interface ImportOptions {
+  /**
+   * The remote URL of the repository to import
+   */
+  sourceRemoteUrl: string;
+  /**
+   * The branch or reference to import
+   */
+  ref: string;
+  /**
+   * The directory in the source repo to import
+   */
+  source: string;
+  /**
+   * The directory in the destination repo to import into
+   */
+  destination: string;
+
+  verbose: boolean;
+  interactive: boolean;
+}
+
+export async function importHandler(options: ImportOptions) {
+  let { sourceRemoteUrl, ref, source, destination } = options;
+
+  output.log({
+    title:
+      'Nx will walk you through the process of importing code from another repository into this workspace:',
+    bodyLines: [
+      `1. Nx will clone the other repository into a temporary directory`,
+      `2. Code to be imported will be moved to the same directory it will be imported into on a temporary branch`,
+      `3. The code will be merged into the current branch in this workspace`,
+      `4. Nx will recommend plugins to integrate tools used in the imported code with Nx`,
+      `5. The code will be successfully imported into this workspace`,
+      '',
+      `Git history will be preserved during this process`,
+    ],
+  });
+
+  const tempImportDirectory = join(tmpdir, 'nx-import');
+
+  if (!sourceRemoteUrl) {
+    sourceRemoteUrl = (
+      await prompt<{ sourceRemoteUrl: string }>([
+        {
+          type: 'input',
+          name: 'sourceRemoteUrl',
+          message:
+            'What is the URL of the repository you want to import? (This can be a local git repository or a git remote URL)',
+          required: true,
+        },
+      ])
+    ).sourceRemoteUrl;
+  }
+
+  try {
+    const maybeLocalDirectory = await stat(sourceRemoteUrl);
+    if (maybeLocalDirectory.isDirectory()) {
+      sourceRemoteUrl = resolve(sourceRemoteUrl);
+    }
+  } catch (e) {
+    // It's a remote url
+  }
+
+  const sourceRepoPath = join(tempImportDirectory, 'repo');
+  const spinner = createSpinner(
+    `Cloning ${sourceRemoteUrl} into a temporary directory: ${sourceRepoPath}`
+  ).start();
+  try {
+    await rm(tempImportDirectory, { recursive: true });
+  } catch {}
+  await mkdir(tempImportDirectory, { recursive: true });
+
+  const sourceGitClient = await cloneFromUpstream(
```

Comment:
If this fails, we may want to ask user to check their URL / directory path. If there is a typo, or what they are importing isn't a git repo.